### ISSUE TITLE
OutOfMemoryError when indexing projects

### ISSUE THREAD

**Issue by michaeljmcd** [oracle/opengrok#1082]
I am running OpenGrok on Windows 7, with Java 8 and Tomcat 8 and I consistently hit an OOM when attempting to index about 170 projects.

```
cd C:\path\to\opengrok-0.12.1.5

SET OPENGROK_TOMCAT_BASE=C:\path\to\apache-tomcat-8.0.26
SET OPENGROK_CONFIGURATION=C:\path\to\opengrok-0.12.1.5\configuration.xml
SET OPENGROK_VERBOSE=true
SET _JAVA_OPTIONS=%JAVA_OPTIONS% -Xmx8g -XX:-UseGCOverheadLimit

java -d64 -server -jar .\lib\opengrok.jar -c C:\path\to\ctags58\ctags.exe -P -s
C:\path\to\src -d C:\path\to\opengrok-0.12.1.5\data\ -W
C:\path\to\configuration.xml -i target -i build -i
tags -i classic.json -i modern.json -i bootstrap.json -i *.zip -i *.tbz2 -i
*.png -i *.jpg -i *.gif -i *.pdf -i *.jar -i *.dll -i *.war -i *.ear -i .idea -i
.settings -i .classpath -i *.exe -C -v -m 8 -r off
```

The machine in question has 16 GB of RAM, so allocating 8GB seems reasonable.

Invariably, the process ends with: 

```
INFO: Progress: 26,358 (99.977%)
Apr 04, 2016 3:14:34 PM org.opensolaris.opengrok.index.IndexDatabase indexDown
INFO: Progress: 26,359 (99.981%)
Apr 04, 2016 3:14:35 PM org.opensolaris.opengrok.index.IndexDatabase indexDown
INFO: Progress: 26,360 (99.985%)
Apr 04, 2016 3:14:35 PM org.opensolaris.opengrok.index.IndexDatabase indexDown
INFO: Progress: 26,361 (99.989%)
Apr 04, 2016 3:14:35 PM org.opensolaris.opengrok.index.IndexDatabase indexDown
INFO: Progress: 26,362 (99.992%)
Apr 04, 2016 3:14:35 PM org.opensolaris.opengrok.index.IndexDatabase indexDown
INFO: Progress: 26,363 (99.996%)
Apr 04, 2016 3:14:35 PM org.opensolaris.opengrok.index.IndexDatabase indexDown
INFO: Progress: 26,364 (100%)
Apr 04, 2016 3:31:59 PM org.opensolaris.opengrok.index.IndexDatabase$1 run
SEVERE: Problem updating lucene index database:
java.lang.OutOfMemoryError: Java heap space
        at java.util.Arrays.copyOf(Arrays.java:3332)
        at java.lang.AbstractStringBuilder.expandCapacity(AbstractStringBuilder.java:137)
        at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:121)
        at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:569)
        at java.lang.StringBuffer.append(StringBuffer.java:369)
        at java.io.BufferedReader.readLine(BufferedReader.java:370)
        at java.io.BufferedReader.readLine(BufferedReader.java:389)
        at org.opensolaris.opengrok.analysis.Ctags.readTags(Ctags.java:182)
        at org.opensolaris.opengrok.analysis.Ctags.doCtags(Ctags.java:173)
        at org.opensolaris.opengrok.analysis.plain.PlainAnalyzer.analyze(PlainAnalyzer.java:78)
        at org.opensolaris.opengrok.analysis.plain.AbstractSourceCodeAnalyzer.analyze(AbstractSourceCodeAnalyzer.java:69)
        at org.opensolaris.opengrok.analysis.AnalyzerGuru.populateDocument(AnalyzerGuru.java:307)
        at org.opensolaris.opengrok.index.IndexDatabase.addFile(IndexDatabase.java:606)
        at org.opensolaris.opengrok.index.IndexDatabase.indexDown(IndexDatabase.java:870)
        at org.opensolaris.opengrok.index.IndexDatabase.indexDown(IndexDatabase.java:835)
        at org.opensolaris.opengrok.index.IndexDatabase.indexDown(IndexDatabase.java:835)
        at org.opensolaris.opengrok.index.IndexDatabase.indexDown(IndexDatabase.java:835)
        at org.opensolaris.opengrok.index.IndexDatabase.indexDown(IndexDatabase.java:835)
        at org.opensolaris.opengrok.index.IndexDatabase.indexDown(IndexDatabase.java:835)
        at org.opensolaris.opengrok.index.IndexDatabase.indexDown(IndexDatabase.java:835)
        at org.opensolaris.opengrok.index.IndexDatabase.indexDown(IndexDatabase.java:835)
        at org.opensolaris.opengrok.index.IndexDatabase.indexDown(IndexDatabase.java:835)
        at org.opensolaris.opengrok.index.IndexDatabase.update(IndexDatabase.java:383)
        at org.opensolaris.opengrok.index.IndexDatabase$1.run(IndexDatabase.java:168)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)

Apr 04, 2016 3:31:59 PM org.opensolaris.opengrok.util.Statistics report
INFO: Done indexing data of all repositories (took 0:19:53)
Apr 04, 2016 3:31:59 PM org.opensolaris.opengrok.util.Statistics report
INFO: Total time: 0:19:53
Apr 04, 2016 3:31:59 PM org.opensolaris.opengrok.util.Statistics report
INFO: Final Memory: 13M/7,165M
```

I have spent some time fiddling with the options based on https://github.com/OpenGrok/OpenGrok/issues/852 but have yet to find anything that will allow the process to complete.

My Java version is shown below.

```
 >java -version
 Picked up _JAVA_OPTIONS:  -Xmx8g -XX:-UseGCOverheadLimit
 java version "1.8.0_45"
 Java(TM) SE Runtime Environment (build 1.8.0_45-b15)
 Java HotSpot(TM) 64-Bit Server VM (build 25.45-b02, mixed mode)
```

Some of the individual source files are fairly large (there are some COBOL programs that are around 3MB a piece, but this does not seem untenable and the problem does not go away when the project containing them is removed), but nothing outlandish.

Is there anything else I can do to get past these issues?

**Comment by tarzanek** [comment]
More detailed logs will have the file name that causes this. 
Note that ctags buffer overflows whole opengrok.
I added fixes for this in trunk build also universal ctags got fixes for this prob.
Unfortunately both versions don't have a released version. 
On solaris/debian you could have build both easily, on windows at least trunk build of opengrok will help you.

Or figure out from log what file exactly causes ctags to misbehave and get it to ignore list.

**Comment by tarzanek** [comment]
Detailed logs - check Opengrok script on how to pass logger properties to jar call ...
I am wondering if with Windows 10 and Ubuntu integration we will now support win through OpenGrok so all systems will be supported ... we might be close :)

**Comment by michaeljmcd** [comment]
so I added the logging configuration from the OpenGrok script to the batch file, but the logs don't really show anything that wasn't there before, just in a different format.

    2016-04-07 09:14:50.976-0500 INFO t20 IndexDatabase.indexDown: Progress: 26,363 (99.996%)
    2016-04-07 09:14:50.977-0500 INFO t20 IndexDatabase.indexDown: Progress: 26,364 (100%)
    2016-04-07 09:29:31.827-0500 SEVERE t15 IndexDatabase$1.run: Problem updating lucene index database: 
    java.lang.OutOfMemoryError: Java heap space
        at java.util.Arrays.copyOf(Arrays.java:3332)
        at java.lang.AbstractStringBuilder.expandCapacity(AbstractStringBuilder.java:137)
        at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:121)
        at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:569)
        at java.lang.StringBuffer.append(StringBuffer.java:369)
        at java.io.BufferedReader.readLine(BufferedReader.java:370)
        at java.io.BufferedReader.readLine(BufferedReader.java:389)
        at org.opensolaris.opengrok.analysis.Ctags.readTags(Ctags.java:182)
        at org.opensolaris.opengrok.analysis.Ctags.doCtags(Ctags.java:173)
        at org.opensolaris.opengrok.analysis.plain.PlainAnalyzer.analyze(PlainAnalyzer.java:78)
        at org.opensolaris.opengrok.analysis.plain.AbstractSourceCodeAnalyzer.analyze(AbstractSourceCodeAnalyzer.java:69)
        at org.opensolaris.opengrok.analysis.AnalyzerGuru.populateDocument(AnalyzerGuru.java:307)
        at org.opensolaris.opengrok.index.IndexDatabase.addFile(IndexDatabase.java:606)
        at org.opensolaris.opengrok.index.IndexDatabase.indexDown(IndexDatabase.java:870)
        at org.opensolaris.opengrok.index.IndexDatabase.indexDown(IndexDatabase.java:835)
        at org.opensolaris.opengrok.index.IndexDatabase.indexDown(IndexDatabase.java:835)
        at org.opensolaris.opengrok.index.IndexDatabase.indexDown(IndexDatabase.java:835)
        at org.opensolaris.opengrok.index.IndexDatabase.indexDown(IndexDatabase.java:835)
        at org.opensolaris.opengrok.index.IndexDatabase.indexDown(IndexDatabase.java:835)
        at org.opensolaris.opengrok.index.IndexDatabase.indexDown(IndexDatabase.java:835)
        at org.opensolaris.opengrok.index.IndexDatabase.indexDown(IndexDatabase.java:835)
        at org.opensolaris.opengrok.index.IndexDatabase.indexDown(IndexDatabase.java:835)
        at org.opensolaris.opengrok.index.IndexDatabase.update(IndexDatabase.java:383)
        at org.opensolaris.opengrok.index.IndexDatabase$1.run(IndexDatabase.java:168)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
    2016-04-07 09:29:31.835-0500 INFO t1 Statistics.report: Done indexing data of all repositories (took 0:16:36)
    2016-04-07 09:29:31.836-0500 INFO t1 Statistics.report: Total time: 0:16:37
    2016-04-07 09:29:31.886-0500 INFO t1 Statistics.report: Final Memory: 14M/7,322M

The verbosity switch is provided and the logging level is `ALL`. I guess I will try compiling from trunk next, unless I am missing something about how to make the logs more detailed. The logging configuration is the standard one with just the paths tweaked:

    .level= ALL

    java.util.logging.FileHandler.pattern = C:/path/to/opengrok-0.12.1.5/opengrok%g.%u.log
    java.util.logging.FileHandler.limit = 52428800
    java.util.logging.FileHandler.count = 3
    java.util.logging.FileHandler.level = ALL
    java.util.logging.FileHandler.formatter = org.opensolaris.opengrok.SimpleFileLogFormatter

**Comment by tarzanek** [comment]
logs should show what analyzer for what file is used and the file that was indexed before crash should be the one responsible ... can you check the log file ?
What version of grok is this? 0.12 or trunk ?
(in trunk this logging might be broken, I didn't verify it since we refactored logging :( )

**Comment by michaeljmcd** [comment]
It is 0.12.1.5 - the version in the file downloads.

**Comment by tarzanek** [comment]
is there any chance you can go and build trunk source and try with trunk? 
Just netbeans and recent jdk8 should get you going ...
(if this is not a production server of course, since trunk has some problems that we need to fix before release (but all of them should go unnoticed by you, they are prob just for huge repos))
I fixed and limited ctags parsing in trunk, so your problems will be history ...
If this is production, then flipping over to linux/solars and using OpenGrok script might yield better debug info - also progress bar slows you down ...
the logging definitely should log file name, alas with your unsupported config and manual run of the jar file we will just troubleshoot unsupported config :(
So basically all these problems are known, in 0.12 we solve them by ignore (also your ignore patterns have " * " in them, which is useless I think), in 0.13 all of them are solved
and on Windows we don't have a proper way to run OpenGrok fully, but I hope this will be lifted once Windows will support linux shells ... (which seems close)
You could even try in a virtual machine I guess, Ubuntu is very easy to install.

**Comment by ahornace** [comment]
Closing due to inactivity. Please try the newest version a please open a new issue if the problem persists.
